### DISCUSSION TITLE
storefront not loading on localhost 3000

### DISCUSSION THREAD

**Discussion by TommyJackson85** [saleor/saleor#7211] [General]
Following inital setup of Saleor and then running ```docker-compose up``` it sets up the dashboard and graphql playground but not the storefront. Is there something I am missing? I am working with version 2.11 This is a screenshot of the terminal showing the storefront setup.

<img width="804" alt="Screenshot 2021-04-20 at 22 35 10" src="https://user-images.githubusercontent.com/22279028/115467165-fbb0ed80-a228-11eb-8265-9ac7de32a742.png">

**Comment by opticrew** [comment]
I have the same problem

**Comment by karolkielecki** [comment]
We've checked it and it's working on our side. Complied with warnings shouldn't be the problem here. Are you working with Docker Desktop on Windows? Sometimes I need to stop running the API and then it works well.

**Comment by krzysztofwolski** [comment]
Is there anything loading at localhost:3000? What's in the console/networking tab of the browser? Initial render takes a while

**Comment by karolkielecki** [comment]
Hi guys! Please try [the following solution](https://spectrum.chat/saleor/saleor-storefront/localhost-3000-not-working-on-initial-setup~7eb8da70-739b-40f7-b148-c80d08f121c5?m=MTYxOTU5NTk2Njg1NQ==)

**Comment by gorankarlic** [comment]
Same issue on MacOS, Saleor v2.11, no response from storefront, dashboard works fine.

```sh
~ > curl -v localhost:3000
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 3000 (#0)
> GET / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.64.1
> Accept: */*
> 
* Empty reply from server
* Connection #0 to host localhost left intact
curl: (52) Empty reply from server
* Closing connection 0
```

```sh
~ > docker -v
Docker version 20.10.5, build 55c4c88
~ > docker-compose -v
docker-compose version 1.29.0, build 07737305
```

**Reply by TommyJackson85** [reply]
i manged to fix it. but i cant remember how haha! I think i didnt setup docker properly or the .env file.

**Reply by opticrew** [reply]
Can you explain how you fix it, please ?

**Reply by gorankarlic** [reply]
I was able to fix the issue by partially reverting saleor-platfrom merge [86](https://github.com/mirumee/saleor-platform/pull/86) which was merged prematurely: contributor states that it "Should be merged when the storefront submodule will be updated with the latest nextjs integration." but latter _has not been merged yet_.

More specifically, the [change](https://github.com/mirumee/saleor-platform/pull/86/commits/c359f7df277617cefbdadcf567394116cbee5b1a#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3L42) in `docker-compose.yml`:

```yaml
command: npm start -- --hostname 0.0.0.0
```

should be reverted to

```yaml
command: npm start -- --host 0.0.0.0
```

for the frontend to work (otherwise the frontend it is only listening to the containers localhost instead of all IPs and is thus innacessible).

**Reply by ilievpavel** [reply]
@gorankarlic MVP 🔝

**Comment by pieceofmine** [comment]
Hello, please I need all your help for running storefront view on port 3000.
Even I modified yml file about command '--host 0.0.0.0' following above guide, storefront is not running still.
(port 3000 is not connected on browser)

Please help me what I have to do more. 
Thanks a lot !!! 


![image](https://user-images.githubusercontent.com/17973652/136929512-238dd365-306c-4706-9da3-49c10857bd44.png)

![image](https://user-images.githubusercontent.com/17973652/136929810-04db53c5-ae32-42a8-a3c4-d774879ad7c9.png)

![image](https://user-images.githubusercontent.com/17973652/136929898-b22bf712-d58f-4ceb-99a2-53595b78fab9.png)

**Reply by Diego17ccs** [reply]
All you have to do is modify the package.json file on line # 199, which should look like this: "start": "cross-env NODE_ENV = develop webpack-dev-server --history-api -fallback --watch --host 0.0.0.0 --port 3000 --mode development --hotOnly ", the only thing you should add is --host 0.0.0.0 after --watch, of course I assume that you are using version 2.11, any questions do not hesitate to ask

**Reply by ivanakcheurov** [reply]
Nope, tried your fix and it didn't help. Same problem appeared (v 2.11): Dashboard and all other services work except storefront. 
I can do a Google Meet call with screen sharing if needed. Otherwise can provide all info needed here.

**Reply by bulentkucuk** [reply]
The same problem happens for me too on Mac.

**Reply by simonwhitaker** [reply]
The docker-compose.yml file has been tweaked to work with storefronts running v3 and above (https://github.com/saleor/saleor-platform/commit/9549f2f13e7158f229d5459fd5ca93c803c71c41). You can revert that commit to get the v2.11 storefront working again.

**Reply by rshanfl2020** [reply]
`  storefront:
    build:
      context: ./saleor-storefront
      dockerfile: ./Dockerfile.dev
    ports:
      - 3000:3000
    restart: unless-stopped
    volumes:
      - ./saleor-storefront/:/app:cached
      - /app/node_modules/
    command: npm start -- --host 0.0.0.0`
This helped me, basically as @simonwhitaker mentioned

**Reply by bulentkucuk** [reply]
@simonwhitaker -- Thanks this has helped.
I'm hoping this issue will be resolved in the next Saleor release, so there won't be a need to revert.

**Reply by simonwhitaker** [reply]
The core of the issue is that in https://github.com/saleor/saleor-platform/commit/9549f2f13e7158f229d5459fd5ca93c803c71c41, the `docker-compose.yml` file was changed to use host networking:

```
network_mode: "host"
```

But this [only works on Linux](https://docs.docker.com/network/host/):

> The host networking driver only works on Linux hosts, and is not supported on Docker Desktop for Mac, Docker Desktop for Windows, or Docker EE for Windows Server.

If, like me, you're working on a Mac (or on Windows), this will be a problem for you.

I think the reason the team have switched to that option in docker-compose.yml is that, in v3, the `saleor-storefront` repo uses NextJs SSR (see release notes [here](https://github.com/saleor/saleor/releases/tag/3.0.0-b.1)). I'm not too familiar with SSR, but I think the gist of it is that, when running in development mode, some content is generated on the client (i.e. the browser, via Javascript), and some content is rendered on the NextJs server. In the case of the `saleor-platform` repo, which uses Docker containers to run the JS dev servers for storefront and dashboard, this means that both the server (inside the `storefront` docker container) and the client (your web browser) need access to the API server (running in the `api` docker container), and using host-mode networking is one way to get this working.

Another option might be to have a separate environment variable (e.g. `SSR_API_URI`) that specifies the API URI for the NextJs server to use when doing SSR. This would default to the same value as `API_URI`, but could be set separately. Then you might be able to get this working cross-platform by updating the SSR code to use this env variable ([here](https://github.com/saleor/saleor-storefront/blob/619024c9d4b3a8e43f6edf8d4958b85bf1b8ac57/src/%40next/utils/ssr.ts#L34), I think) changing the docker-compose.yml file to something like this:

1. Remove the `network-mode: host` line
2. Add `networks: [ saleor-backend-tier ]` to the `storefront` service
3. Set environment variables in the `storefront` service, along the lines of:
    - `API_URI=http://localhost:8000/graphql/`
    - `SSR_API_URI=http://api:8000/graphql/`

I haven't tried this out, it's just a suggestion for a possible direction that keeps this working cross-platform in a post-3.0 world.

**Comment by lionelquellery** [comment]
On my side, I have remove "network_mode: host" and I have added "ports: - 3000:3000" it looks like that:
 : 
` storefront:
    build:
      context: ./saleor-storefront
      dockerfile: ./Dockerfile.dev
    restart: unless-stopped
    ports:
      - 3000:3000
    # network_mode: "host"
    volumes:
      - ./saleor-storefront/:/app:cached
      - /app/node_modules/
    command: npm start -- --host 0.0.0.0
`

After this it works well and I have had the storefront.
